src/main/java/im/status/keycard/SecureChannel.java

```diff
@@ -207,33 +290,60 @@ public void pair(APDU apdu) {
   }
 
   /**
-   * Performs the first step of pairing. In this step the card solves the challenge sent by the client, thus authenticating
-   * itself to the client. At the same time, it creates a challenge for the client. This can only fail if the card has
-   * already paired with the maximum allowed amount of clients.
+   * Performs the first step of certificate based pairing. In this step, the card provides a CA signed certificate
+   * of its card ID key, and proves ownership of this key with a signature on a challenge hash. The challenge hash
+   * to be signed is computed as the sha256 hash of the client salt (provided by client) and card salt (generated randomly).
+   * The card will also include its random salt in the response to the client, so that the client can reproduce the challenge
+   * hash and verify the card signature.
    *
    * @param apduBuffer the APDU buffer
    * @return the length of the reply
    */
   private short pairStep1(byte[] apduBuffer) {
-    preassignedPairingOffset = -1;
-
-    for (short i = 0; i < (short) pairingKeys.length; i += PAIRING_KEY_LENGTH) {
-      if (pairingKeys[i] == 0) {
-        preassignedPairingOffset = i;
-        break;
-      }
+    // Make sure certificate exisits
+    if (idCertStatus != ID_CERTIFICATE_LOCKED) {
+      ISOException.throwIt(ISO7816.SW_CONDITIONS_NOT_SATISFIED);
     }
 
-    if (preassignedPairingOffset == -1) {
-      ISOException.throwIt(ISO7816.SW_FILE_FULL);
+    // Override first pairing slot
+    preassignedPairingOffset = 0;
+
+    // Compute ECDH secret
+    crypto.ecdh.init(idKeypair.getPrivate());
+    short pubKeyOff = (short)(ISO7816.OFFSET_CDATA + SC_SECRET_LENGTH);
+    try {
+      crypto.ecdh.generateSecret(apduBuffer, (short) (pubKeyOff + 1), apduBuffer[pubKeyOff], secret, (short) 0);
```

Comment:
Just FYI, `secret` is still defined as being 64 bytes. Not sure if status is using it for something that requires that much space, but our pairing process only uses 32 bytes.

Not a big deal either way, just preferable to keep things as small as possible

Reply:
Yes, it needs to be 64 bytes. The secure channel keys are derived from two halves of a sha512. This buffer holds both keys in that step